File: WdtcCore/src/main/java/org/wdt/wdtc/core/manger/SettingManger.java
package org.wdt.wdtc.core.manger;


import lombok.Data;
import lombok.SneakyThrows;
import lombok.experimental.Accessors;
import org.apache.log4j.Logger;
import org.wdt.utils.gson.Json;
import org.wdt.utils.gson.JsonUtils;
import org.wdt.wdtc.core.utils.JavaUtils;
import org.wdt.wdtc.core.utils.WdtcLogger;

import java.io.File;
import java.io.IOException;
import java.util.HashSet;
import java.util.Set;

public class SettingManger {
  private static final Logger logmaker = WdtcLogger.getLogger(SettingManger.class);


  @SneakyThrows(IOException.class)
  public static Setting getSetting() {
    return JsonUtils.readFileToClass(FileManger.getSettingFile(), Setting.class);
  }


  @SneakyThrows
  public static void putSettingToFile(Setting setting) {
    JsonUtils.writeObjectToFile(FileManger.getSettingFile(), setting, Json.getBuilder().setPrettyPrinting());
  }

  @Data
  @Accessors(chain = true)
  public static class Setting {
    private final boolean isConsole = VMManger.isConsole();
    private DownloadSourceManger.DownloadSourceList DownloadSource = DownloadSourceManger.DownloadSourceList.OFFICIAL;
    private boolean Console = false;
    private boolean LlvmpipeLoader = false;
    private File DefaultGamePath = new File(System.getProperty("user.dir"));
    private Set<JavaUtils.JavaInfo> JavaPath = new HashSet<>();
    private boolean ChineseLanguage = true;
    private double WindowsWidth = 616.0;
    private double WindowsHeight = 489.0;
    private String PreferredVersion;
  }
}

File: WdtcCore/src/main/kotlin/org/wdt/wdtc/core/manger/SettingManger.kt
@file:JvmName("SettingManger")

package org.wdt.wdtc.core.manger

import org.wdt.utils.gson.readFileToClass
import org.wdt.utils.gson.writeObjectToFile
import org.wdt.wdtc.core.game.Version
import org.wdt.wdtc.core.utils.JavaUtils.JavaInfo
import org.wdt.wdtc.core.utils.gson.serializeVersionGson
import java.io.File

data class Setting(
	var downloadSource: DownloadSourceKind = DownloadSourceKind.OFFICIAL,
	var console: Boolean = false,
	var llvmpipeLoader: Boolean = false,
	var defaultGamePath: File = File(System.getProperty("user.dir")),
	var javaPath: MutableSet<JavaInfo> = HashSet(),
	var chineseLanguage: Boolean = true,
	var windowsWidth: Double = 616.0,
	var windowsHeight: Double = 489.0,
	var preferredVersion: Version? = null
)


val currentSetting: Setting
	get() = settingFile.readFileToClass(serializeVersionGson)

fun Setting.putSettingToFile() {
	settingFile.writeObjectToFile(this, serializeVersionGson)
}

inline fun Setting.changeSettingToFile(block: Setting.() -> Unit): Setting {
	return apply(block).also { it.putSettingToFile() }
}